### PR TITLE
R2 bucket abstraction

### PR DESCRIPTION
**Pull request by AntonCroner**

### CODE REVIEW

File: Bucket/Blob.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	async remove(key: string | string[]): Promise<void> {
+		this.backend.delete(key)
+	}
+	async list(options?: Bucket.ListOptions): Promise<(Partial<Blob.Result<M>> & { key: string })[]> {
+		const list = (await this.backend.list(options)).objects.map(i => ({
+			key: i.key,
+			meta: i.customMetadata as M,
+		}))
+		return options?.values
+			? await Promise.all(list.map(i => ({ ...i, value: this.get(i.key).then(v => v?.value) })))
```
Comment (EliasEriksson) [inline Bucket/Blob.ts:24]: the promise behind the `value` property should be awaited?
running this in the javascript console results in the value being promises in the objects
```js
await Promise.all([1, 2, 3].map(async (value) => ({value: new Promise(resolve => resolve(value))}))) 
```
image

File: Bucket/Blob.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import * as platform from "@cloudflare/workers-types"
+import { Bucket } from "./Bucket"
+
+export class Blob<M extends Record<string, string> | undefined = undefined> extends Bucket<platform.Blob, M> {
+	async store(key: string, value: platform.Blob, meta: M): Promise<Blob.Result<M> | undefined> {
+		await this.backend.put(key, value, { customMetadata: meta })
```
Comment (EliasEriksson) [inline Bucket/Blob.ts:6]:
```ts
await this.backend.put(key, value, { customMetadata: meta })
```
might need to be turned into 
```ts
await this.backend.put(key, value.stream(), { customMetadata: meta })
```

**Review by EliasEriksson — DISMISSED** [review]

File: Bucket/Blob.ts (hunk 1 above)
Comment (AntonCroner) [reply Bucket/Blob.ts]: Fixed.

File: Bucket/Blob.ts (hunk 2 above)
Comment (AntonCroner) [reply Bucket/Blob.ts]: Fixed.

**Review by fredrikbryntesson — APPROVED** [review]

**Pull request merged**